src/partmc.F90

```diff
@@ -454,279 +450,37 @@ subroutine partmc_part(file)
-       max_buffer_size = max_buffer_size &
-            + pmc_mpi_pack_size_scenario(scenario)
-       max_buffer_size = max_buffer_size &
-            + pmc_mpi_pack_size_env_state(env_state_init)
-       max_buffer_size = max_buffer_size &
-            + pmc_mpi_pack_size_integer(rand_init)
-       max_buffer_size = max_buffer_size &
-            + pmc_mpi_pack_size_logical(do_restart)
-       max_buffer_size = max_buffer_size &
-            + pmc_mpi_pack_size_logical(do_init_equilibrate)
-       max_buffer_size = max_buffer_size &
-            + pmc_mpi_pack_size_aero_state(aero_state_init)
-
-       allocate(buffer(max_buffer_size))
-
-       position = 0
-       call pmc_mpi_pack_run_part_opt(buffer, position, run_part_opt)
-       call pmc_mpi_pack_real(buffer, position, n_part)
-       call pmc_mpi_pack_gas_data(buffer, position, gas_data)
-       call pmc_mpi_pack_gas_state(buffer, position, gas_state_init)
-       call pmc_mpi_pack_aero_data(buffer, position, aero_data)
-       call pmc_mpi_pack_aero_dist(buffer, position, aero_dist_init)
-       call pmc_mpi_pack_scenario(buffer, position, scenario)
-       call pmc_mpi_pack_env_state(buffer, position, env_state_init)
-       call pmc_mpi_pack_integer(buffer, position, rand_init)
-       call pmc_mpi_pack_logical(buffer, position, do_restart)
-       call pmc_mpi_pack_logical(buffer, position, do_init_equilibrate)
-       call pmc_mpi_pack_aero_state(buffer, position, aero_state_init)
-       call assert(181905491, position <= max_buffer_size)
-       buffer_size = position ! might be less than we allocated
-    end if
-
-    ! tell everyone the size
-    call pmc_mpi_bcast_integer(buffer_size)
-
-    if (pmc_mpi_rank() /= 0) then
-       ! non-root processes allocate space
-       allocate(buffer(buffer_size))
-    end if
-
-    ! broadcast data to everyone
-    call pmc_mpi_bcast_packed(buffer)
-
-    if (pmc_mpi_rank() /= 0) then
-       ! non-root processes unpack data
-       position = 0
-       call pmc_mpi_unpack_run_part_opt(buffer, position, run_part_opt)
-       call pmc_mpi_unpack_real(buffer, position, n_part)
-       call pmc_mpi_unpack_gas_data(buffer, position, gas_data)
-       call pmc_mpi_unpack_gas_state(buffer, position, gas_state_init)
-       call pmc_mpi_unpack_aero_data(buffer, position, aero_data)
-       call pmc_mpi_unpack_aero_dist(buffer, position, aero_dist_init)
-       call pmc_mpi_unpack_scenario(buffer, position, scenario)
-       call pmc_mpi_unpack_env_state(buffer, position, env_state_init)
-       call pmc_mpi_unpack_integer(buffer, position, rand_init)
-       call pmc_mpi_unpack_logical(buffer, position, do_restart)
-       call pmc_mpi_unpack_logical(buffer, position, do_init_equilibrate)
-       call pmc_mpi_unpack_aero_state(buffer, position, aero_state_init)
-       call assert(143770146, position == buffer_size)
+       if (.not. do_restart) then
+          call uuid4_str(run_part_opt%uuid)
+       end if
```

Comment:
I think we can combine all of the above code into a single `if (pmc_mpi_rank() == 0)` block? That is, something like the following. I think this is ok because the only reason that we initialize the RNG is for UUID generation, which only happens on rank zero, so it's clearer to group this code together when it's really needed.
```
if (pmc_mpi_rank() == 0) then
    call spec_file_read_run_part(...)
    if (.not. do_restart) then
        ! initialize RNG with random seed for UUID generation
        call pmc_srand(...)
        call uuid4_str(...)
        call pmc_rand_finalize(...)
    end if
end if
```

Reply:
I think the issue is that there is a MPI broadcast in `pmc_srand`. Therefore I think we currently need to initialize everyone even if only the root does anything.

Reply:
Looking at this, I think we run into an issue here if not all ranks call `pmc_srand()` because there is a `pmc_mpi_bcast_integer()` in that subroutine.
https://github.com/compdyn/partmc/blob/1b4742ef850803d95ece31f649efb760f3cdb2f7/src/rand.F90#L90-L99

We could either leave it as is or pass an optional argument to `pmc_srand()` to avoid doing this broadcast when we are calling just with rank-0 and do what you proposed.

Reply:
Good point. Let's not bother changing that. In this case I guess I'd do something like:
```
! initialize RNG with random seed for UUID generation
call pmc_srand(...)

if (pmc_mpi_rank() == 0) then
    call spec_file_read_run_part(...)
    if (.not. do_restart) then
        call uuid4_str(...)
    end if
end if

! re-initialize RNG with the given seed
call pmc_rand_finalize()
call pmc_srand(rand_init, pmc_mpi_rank())
```

Reply:
Alright - https://github.com/compdyn/partmc/pull/188/commits/71afc15f3ebe193c9d8c595a81b3bd287b386ffd makes this cleaner.